### PR TITLE
Add Q computation to utils/math

### LINKED ISSUE #7671: Add Q96 computation in lisk-utils/math
### Description

- Add Q96 computation in @liskhq/lisk-utils/math
  - Class Q(val: number | bigint | Buffer, base = 96)
    - add(n: Q): Q
    - sub(n: Q): Q
    - mul(n: Q): Q
    - div(n: Q): Q
    - muldiv(n: Q, m: Q): Q
    - toInt(operation: roundDown | roundUp): bigint
    - Inv(n: Q): Q
    - toBytes(): Buffer

### Acceptance Criteria

- All functions are unit tested

#### Definition

We represent a positive real number `r` as a `Qn` by the positive integer `Qn(r) = floor(r*2^n)` (inspired by the [Q number format](https://en.wikipedia.org/wiki/Q_(number_format)) for signed numbers). In this representation, we have `n` bits of fractional precision. We do not limit the size of `r` as modern libraries can handle integers of arbitrary size; note that all the `Qn` conversions assume no loss of significant digits.

#### Operations on Integers

For an integer `a` in the `Qn` format, we define:

* Rounding down: `roundDown_n(a) = a >> n`

#### Qn Arithmetic Operations

In the following definition, `*` is the integer multiplication, `//` is the integer division (rounded down). Division by zero is not permitted and should raise an error in any implementation.

For two numbers `a`,`b` and `c` in `Qn` format, we define the following arithmetic operations which all return a `Qn`:

* Addition: `add_n(a, b) = a + b`

* Subtraction: `sub_n(a,b) = a - b`, only defined if `a >= b`

* Multiplication: `mul_n(a, b) = (a * b) >> n`

* Division: `div_n(a, b) = (a << n) // b`

* Multiplication and then division: `muldiv_n(a, b, c) = roundDown_n(((a * b) << n) // c)`

* Multiplication and then division, rounding up: `muldiv_n_RoundUp(a, b, c) = roundUp_n(((a * b) << n) // c)`

* Convert to integer rounding down: `Q_n_ToInt(a) = roundDown_n(a)`

* Convert to integer rounding up: `Q_n_ToIntRoundUp(a) = roundUp_n(a)`

* Inversion in the decimal precision space: `inv_n(a) = div_n(1 << n, a)`

### PR DESCRIPTION
**Pull request by shuse2**
### What was the problem?

This PR resolves #7671 

### How was it solved?

- Add Q calculation with Q96 class exposed as well

### How was it tested?

- Test implementation with fixtures

### CODE REVIEW

**Review by sergeyshemyakov — APPROVED** [review]
I have not reviewed the tests with 100% care, I assume they pass successfully. The PR looks good, but I have one comment. 

BTW, Emilio has implemented [his version of Q96 numbers for DEX](https://github.com/LiskHQ/lisk-dex-core/blob/development/src/app/modules/dex/utils/q96.ts), do you know what will be the relation between two implementations?

File: elements/lisk-utils/src/math/q.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+		}
+		this._val = BigInt(`0x${val.toString('hex')}`);
+	}
+
+	public add(n: Q) {
+		if (this._base !== n._base) {
+			throw new Error('Base of input does not match.');
+		}
+		this._val += n._val;
+		return this;
```
[the remark this answers is not in the document]
Comment (shuse2) [reply elements/lisk-utils/src/math/q.ts:43]: Q96 is needed on SDK for DPoS, so we added as part of library.

good point about mutability, since it's number, probably i shouldve kept it as @Incede did. Ill update to be immutable

File: elements/lisk-utils/src/math/q.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { MAX_FRAC, ONE, Q_OPERATION, TWO } from './constants';
+
+const numberToQ = (base: bigint, val: number): bigint => {
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:18]: This function handles conversion of a number to Q notation. Shouldn't it belong inside `Q` class, instead of keeping it as an external function?

File: elements/lisk-utils/src/math/q.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	const denominator = TWO ** base;
+	const int = Math.floor(val);
+	const decimals = val - int;
+	if (decimals === 0) {
+		return BigInt(val) * denominator;
+	}
+	const [_, binaryStr] = decimals.toString(2).split('.');
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:25]: A super tiny remark: variable `_` is not necessary here 🙂

This is a valid JavaScript/TypeScript syntax:

```typescript
const [, binaryStr] = decimals.toString(2).split('.');
```

File: elements/lisk-utils/src/math/q.ts (hunk 2 above)
Comment (shuse2) [reply elements/lisk-utils/src/math/q.ts:18]: this is util function which doesn't use any context.
Also, it's used in the static function, so it cannot be method, so i just preferred to be an util function

File: elements/lisk-utils/src/math/q.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+		}
+		if (typeof val === 'bigint') {
+			return new Q(val * TWO ** bigintBase, bigintBase);
+		}
+		return new Q(BigInt(`0x${val.toString('hex')}`), bigintBase);
+	}
+
+	public add(n: Q) {
+		if (this._base !== n._base) {
+			throw new Error('Base of input does not match.');
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:62]: Instead of copying the same code in every public method, I think it is cleaner to handle that recurring task using a private method, e.g. `checkBase()`

File: elements/lisk-utils/src/math/q.ts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	private readonly _base: bigint;
+	private readonly _val: bigint;
+
+	public constructor(val: bigint, base: number | bigint) {
+		this._base = BigInt(base);
+		if (this._base > MAX_FRAC) {
+			throw new Error('Base exceeds max fraction allowed.');
+		}
+		this._val = val;
+		this._base = BigInt(base);
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:43]: You've already made this assignment just a few lines above.

File: elements/lisk-utils/src/math/q.ts [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	}
+
+	public static fromValue(val: number | bigint | Buffer | Q, base: number | bigint) {
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:46]: Do we really need 1 big method that handles different input formats, where there is no overlapping logic at all?

I would prefer to break this down into 3 smaller methods:

* `fromNumber()`
* `fromBigInt()`
* `fromBuffer()`

Not sure that we need `fromQ()` at all, because we can directly do it using the constructor.

File: elements/lisk-utils/src/math/q.ts [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+const numberToQ = (base: bigint, val: number): bigint => {
+	const denominator = TWO ** base;
+	if (Number.isInteger(val)) {
+		return BigInt(val) * denominator;
+	}
+	const [intStr, fractionalStr] = val.toString().split('.');
+	return (
+		BigInt(intStr) * denominator +
+		(BigInt(fractionalStr) * denominator) / BigInt(10) ** BigInt(fractionalStr.length)
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:26]: We can completely get rid of the `denominator` variable and replace multiplication with `<< base`.

Comment (bobanm) [reply elements/lisk-utils/src/math/q.ts:26]: Don't forget the brackets:

```typescript
const numberToQ = (base: bigint, val: number): bigint => {
	if (Number.isInteger(val)) {
		return BigInt(val) << base;
	}
	const [intStr, fractionalStr] = val.toString().split('.');
	return (
		(BigInt(intStr) << base) +
		(BigInt(fractionalStr) << base) / BigInt(10) ** BigInt(fractionalStr.length)
	);
};
```

File: elements/lisk-utils/src/math/q.ts [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	public inv() {
+		const numerator = new Q(ONE << this._base, this._base);
+		return numerator.div(this);
+	}
+
+	public toInt(operation: Q_OPERATION = Q_OPERATION.ROUND_DOWN) {
+		if (operation === Q_OPERATION.ROUND_DOWN) {
+			return this._roundDown(this._val);
+		}
+		return this._roundUp(this._val);
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:111]: This looks awkward.

I don't think we need `toInt()` method. I propose to be consistent with JavaScript math functions and have `ceil()` and `floor()` instead. That could be what `_roundDown()` and `roundUp()` are all about.

File: elements/lisk-utils/src/math/q.ts [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+	public gt(n: Q): boolean {
+		return this._val > n._val;
+	}
+
+	public gte(n: Q): boolean {
+		return this._val >= n._val;
+	}
+
+	private _roundDown(n: bigint) {
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:138]: Rename to `floor()` and make public?

File: elements/lisk-utils/src/math/q.ts [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+		return n >> this._base;
+	}
+
+	private _roundUp(n: bigint) {
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:142]: Rename to `ceil()` and make public?

File: elements/lisk-utils/src/math/q.ts [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+		const x = ONE << this._base;
+		const y = n % x;
+
+		if (y === BigInt(0)) {
+			return n >> this._base;
+		}
+
+		const r = n >> this._base;
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:150]: `x`, `y` and `r` are very poor choices for variable names

File: elements/lisk-utils/src/math/q.ts [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+		return new Q((this._val << this._base) / n._val, this._base);
+	}
+
+	public muldiv(n: Q, m: Q) {
+		if (this._base !== n._base || this._base !== m._base) {
+			throw new Error('Invalid input, base does not match.');
+		}
+		const x = this._val * n._val;
+		const y = x << this._base;
+		const z = y / m._val;
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:97]: `x`, `y` and `z` are very poor choices for variable names

File: elements/lisk-utils/src/math/q.ts (hunk 6 above)
Comment (shuse2) [reply elements/lisk-utils/src/math/q.ts:46]: in utils, if we accept different values, anyway it will require to check the type of the input. This is factory method so i think it's fine to receive different types.
the usage will be like
```
q96(3).add(q96(3))
```
instead of
```
Q.fromNumber(3, 96).add(Q.fromNumber(3, 96))
```

File: elements/lisk-utils/src/math/q.ts (hunk 9 above)
Comment (shuse2) [reply elements/lisk-utils/src/math/q.ts:138]: Remove the internal usage and changed it to floor

File: elements/lisk-utils/test/math/q.spec.ts [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+		const numberCases = [
+			[{ input: 0, expected: expectedQ(base, 0), binary: '0' }],
+			[{ input: 10, expected: expectedQ(base, 10), binary: '1010' }],
+			[{ input: 10.25, expected: expectedQ(base, 10, [2]), binary: '1010.01' }],
+			[{ input: 0.25, expected: expectedQ(base, 0, [2]), binary: '0.01' }],
+			[{ input: 0.23, expected: expectedQ(base, 0, [3, 4, 5, 7]), binary: '0.0011101011' }],
+		];
+		it.each(numberCases)('should create Q from number', val => {
+			const actual = q(val.input, base);
+			expect(actual.toBuffer().toString('hex')).toEqual(bigintToHex(val.expected));
```
Comment (bobanm) [inline elements/lisk-utils/test/math/q.spec.ts:47]: This assertion looks very cumbersome.

What's the reason to compare strings, instead of exposing q value bigints and comparing them?

File: elements/lisk-utils/src/math/q.ts (hunk 9 above)
Comment (shuse2) [reply elements/lisk-utils/src/math/q.ts:138]: Actually the usage is not same as `floor`, internally so ill keep this and expose ceil/floor using those function

File: elements/lisk-utils/test/math/q.spec.ts [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+	// mulBasePow2 computes 2 ** base * val
+	const mulBasePow2 = (b: number, val: number | bigint) => pow2(b) * BigInt(val);
+	// pow2BaseDiffAddition accumulates 2 ** (base-vals[n])
+	const pow2BaseDiffAddition = (b: number, vals: number[]) =>
+		vals.reduce((prev, curr) => prev + pow2(b - curr), BigInt(0));
+	// ExpectedQ generates the Q format big integer
+	// int: integer part of the result
+	// decimalBinaryPosthe elements should bein the array means the position of "1" in binary.
+	const expectedQ = (b: number, int: number, decimalBinaryPos?: number[]) =>
+		mulBasePow2(b, int) + pow2BaseDiffAddition(b, decimalBinaryPos ?? []);
```
Comment (bobanm) [inline elements/lisk-utils/test/math/q.spec.ts:34]: Can we get rid of all this junk by using cleaner test data? 🙂

File: elements/lisk-utils/test/math/q.spec.ts [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+
+import { Q_OPERATION, q, q96 } from '../../src/math';
+
+/**
+ * For the test cases `binary` represents the calculation result in binary.
+ */
+describe('Q', () => {
+	//  In this test Q8 is used as example because 8 bit is easier to generate the tests.
+	// For Q8 after 9th bit should be ignored
+	const base = 8;
```
Comment (bobanm) [inline elements/lisk-utils/test/math/q.spec.ts:23]: I think that the tests would be more relevant if the majority of them would use `base = 96`, with only a few tests using another base. Not the other way around.

That would make our tests more realistic, and also provide some sort of stress testing.

File: elements/lisk-utils/test/math/q.spec.ts [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { Q_OPERATION, q, q96 } from '../../src/math';
+
+/**
+ * For the test cases `binary` represents the calculation result in binary.
```
Comment (bobanm) [inline elements/lisk-utils/test/math/q.spec.ts:18]: Is `binary` value ever used as an input or output of any test, or is it there only to support some manual checks?

It looks like we could get rid of it completely.

File: elements/lisk-utils/test/math/q.spec.ts (hunk 15 above)
Comment (shuse2) [reply elements/lisk-utils/test/math/q.spec.ts:23]: with base 96, it's quite difficult to check the edge case, where the bits are truncated.
Also, there is tests for q96 as well

File: elements/lisk-utils/src/math/q.ts [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+	private _roundUp(n: bigint) {
+		const x = ONE << this._base;
+		const y = n % x;
+
+		if (y === BigInt(0)) {
+			return n >> this._base;
+		}
+
+		const r = n >> this._base;
+		return r + ONE;
```
[the remark this answers is not in the document]
Comment (shuse2) [reply elements/lisk-utils/src/math/q.ts:151]: i dont think the code is correct. we want to return `n >> base` if n is multiply of `2 ** base` otherwise add 1 to the result

[the remark this answers is not in the document]
Comment (shuse2) [reply elements/lisk-utils/src/math/q.ts:151]: ah nvm, I misinterpreted the code

**Review by Incede — APPROVED** [review]
Division by zero needs to be checked for all relevant cases, otherwise LGTM. Also maybe we can create an issue to check the performance in worst case scenarios and optimize later.

File: elements/lisk-utils/test/math/q.spec.ts (hunk 16 above)
Comment (shuse2) [reply elements/lisk-utils/test/math/q.spec.ts:18]: binary is used more like test description.
From the binary result, the exact q value is constructed with 8bit accuracy

File: elements/lisk-utils/test/math/q.spec.ts (hunk 14 above)
Comment (shuse2) [reply elements/lisk-utils/test/math/q.spec.ts:34]: Any suggestion to construct good test data?

File: elements/lisk-utils/test/math/q.spec.ts (hunk 13 above)
Comment (shuse2) [reply elements/lisk-utils/test/math/q.spec.ts]: Updated to check internal value, but usually we want to check the public interface, and exposing internal `_val` does not make sense

File: elements/lisk-utils/src/math/q.ts [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { ONE, Q_OPERATION, TWO, ZERO } from './constants';
+
+const numberToQ = (base: bigint, val: number): bigint => {
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:18]: `Q.constructor()` and `Q.fromValue()` both use `val` as the first and `base` as the second argument.

I think that `numberToQ()` should be consistent with that convention.

File: elements/lisk-utils/src/math/q.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (bobanm) [reply elements/lisk-utils/src/math/q.ts]: You are right. My code handles only the top edge case, but not the bottom edge case. Adding another check would make it not that simple, so let's stick with the good old modulus division 🙂

File: elements/lisk-utils/test/math/q.spec.ts (hunk 16 above)
Comment (bobanm) [reply elements/lisk-utils/test/math/q.spec.ts]: Cleaned up and pushed

File: elements/lisk-utils/test/math/q.spec.ts (hunk 14 above)
Comment (bobanm) [reply elements/lisk-utils/test/math/q.spec.ts]: Cleaned up and pushed 🧹

**Comment by bobanm** [comment]
After a thorough review and update process, here is finally a "LGTM" from me 🙂

You earned this one, @shuse2 🏆

File: elements/lisk-utils/src/math/q.ts [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+			throw new Error('Invalid input, output cannot be negative.');
+		}
+
+		return new Q(this._val - n._val, this._base);
+	}
+
+	public mul(n: Q) {
+		this._checkBase(n);
+
+		return new Q(this._roundDown(this._val * n._val), this._base);
```
Comment (bobanm) [inline elements/lisk-utils/src/math/q.ts:73]: While using `_roundDown()` here is mathematically equivalent as what we need to do, semantically it is very confusing.

To avoid confusion, better use:

```typescript
return new Q(this._val * n._val >> this._base, this._base);
```
